cpp/src/arrow/csv/writer.cc

```diff
@@ -328,6 +359,31 @@ class QuotedColumnPopulator : public ColumnPopulator {
   }
 
  private:
+  // Returns true if there's no quote in the string array
+  // similar to std::find, but with much better performance
+  static bool NoQuoteInArray(const StringArray& array) {
+    const uint8_t* const data = array.raw_data() + array.value_offset(0);
+    const int64_t buffer_size = array.total_values_length();
+    int64_t offset = 0;
+#if defined(ARROW_HAVE_SSE4_2) || defined(ARROW_HAVE_NEON)
+    using simd_batch = xsimd::make_sized_batch_t<uint8_t, 16>;
+    while ((offset + 16) <= buffer_size) {
+      const auto v = simd_batch::load_unaligned(data + offset);
+      if (xsimd::any(v == '"')) {
+        return false;
+      }
+      offset += 16;
+    }
+#endif
+    while (offset < buffer_size) {
+      if (data[offset] == '"') {
```

Comment:
I think it might be the same to use CountQuote here?

Reply:
Will revisit this function, together with the above simd code, to see if we can get a simpler code without performance loss.

Reply:
Refactored.